### PR TITLE
Show QGIS news items on welcome page 

### PR DESCRIPTION
**Pull request by nyalldawson**
The curated QGIS news feed is now shown on the welcome page. This finally gives us a direct channel to push project news to ALL our users!

Peek 2019-07-09 15-10

Some notes:
- Let's not get stuck on the visual details here yet -- these can better be refined post merge. Please review the existing code only. I'm well aware we've got work to do in improving the appearance here, e.g. I'd like to color code the different item types (projects vs templates vs news), but I'd rather discuss individual atomic changes in later PRs. 
- Currently, there's no news service running on the qgis architecture, so to test locally you need an instance of https://github.com/elpaso/qgis-feed and you'll need to modify the hardcoded news feed url to match the instance.
- Please no discussion here about allowing configurable feed urls. This is out of scope for this work as I'm sure this discussion will be lengthy and tricky to handle, and I don't want this PR delayed while we work out if we want to allow this or not. (In the end, it's a one-line change if desirable!). Let's discuss that on the dev mailing list instead.
- Please no discussion here about whether we want a news feed or not -- PSC has already made the call that a (optionally disable-able) news feed is desirable, so that decision is locked in!

### CODE REVIEW

**Comment by nyalldawson** [comment]
Further details on the technical side of things here:
- the news feed is fetched once at startup, via a non-blocking background task. Only items since the last successful fetch are requested from the server. (ie. usually the response will be empty)
- items are cached locally. These are stored directly in QSettings for simplicity and to avoid the complexity of another user database, especially given that at most we'll only ever see 10-20 items here, and generally it'll only be 1 or 2.
- images for items are fetched in the background, and are cached locally. They are only ever fetched once per news item.
- When an item is dismissed, it is removed from the QSettings store and the thumbnail is deleted. This item is NOT refetched on next launch because we send the last successful fetch time in the request -- so old items (whether dismissed or not) are never re-fetched.

File: images/images.qrc [hunk 1]
```diff
@@ -767,6 +767,7 @@
         <file>themes/default/mIconAlignRight.svg</file>
         <file>themes/default/mIconArrangeSymbolsLeft.svg</file>
         <file>themes/default/mIconArrangeSymbolsRight.svg</file>
+        <file>themes/default/mIconClearNews.svg</file>
```
Comment (nirvn) [inline images/images.qrc]: I'd rename that to mIconClearItem.svg , as I'd love to add this icon to recent projects and templates.

File: src/app/qgswelcomepage.cpp [hunk 2]
```diff
@@ -277,9 +340,86 @@ void QgsWelcomePage::showContextMenuForTemplates( QPoint point )
...
+  } );
+  menu->addAction( dismissAction );
+  QAction *dismissAllAction = new QAction( tr( "Dismiss All" ), menu );
+  connect( dismissAllAction, &QAction::triggered, this, [this]
+  {
+    mNewsFeedParser->dismissAll();
+  } );
+  menu->addAction( dismissAllAction );
+  menu->addSeparator();
+  QAction *hideAction = new QAction( tr( "Hide QGIS News…" ), menu );
```
Comment (nirvn) [inline src/app/qgswelcomepage.cpp]: Great integration; we could possibly add similar actions to the recent projects and template items (in a separate PR).

File: src/app/qgswelcomepage.cpp [hunk 3]
```diff
@@ -277,9 +340,86 @@ void QgsWelcomePage::showContextMenuForTemplates( QPoint point )
...
+  connect( hideAction, &QAction::triggered, this, [this]
+  {
+    if ( QMessageBox::question( this,  tr( "QGIS News" ), tr( "Are you sure you want to permanently hide QGIS news? This action can not be undone." ) ) == QMessageBox::Yes )
```
Comment (nirvn) [inline src/app/qgswelcomepage.cpp]: TODO: make this undo-able, in a different PR that'd cover hiding/showing individual welcome page sections in the options dialog. 

Not part of this PR, but might be worth adding the TODO note to remember to change the string here.

**Review by nirvn — APPROVED** [review]
Nice work; quickly looked over the code, nothing jumped at me.

File: src/app/qgswelcomepage.cpp (hunk 3 above)
Comment (elpaso) [reply src/app/qgswelcomepage.cpp]: Or maybe like in QT creator "restore default layout" for the whole welcome page.

File: src/core/qgsnewsfeedparser.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ */
+class CORE_EXPORT QgsNewsFeedParser : public QObject
+{
+    Q_OBJECT
+  public:
+
+    /**
+     * Represents a single entry from a news feed.
+     */
+    class Entry
```
Comment (m-kuhn) [inline src/core/qgsnewsfeedparser.h]: /nitpickmode
Data only classes are better made structs
/nitpickmode off

File: src/core/qgsnewsfeedparser.h [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+     */
+    static QString keyForFeed( const QString &baseUrl );
+
+  public slots:
+
+    /**
+     * Fetches new entries from the feed's URL.
+     * \see fetched()
+     */
+    void fetch();
```
Comment (m-kuhn) [inline src/core/qgsnewsfeedparser.h]: Any reason to make this a slot?

File: src/core/qgsnewsfeedparser.cpp [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+      previewPainter.end();
+
+      // Save image, so we don't have to fetch it next time
+      const QString previewDir = QStringLiteral( "%1/previewImages" ).arg( QgsApplication::qgisSettingsDirPath() );
+      QDir().mkdir( previewDir );
+      const QString imagePath = QStringLiteral( "%1/%2.png" ).arg( previewDir ).arg( entry.key );
+      previewImage.save( imagePath );
+
+      mEntries[ entryIndex ].image = QPixmap::fromImage( previewImage );
+      this->emit imageFetched( entry.key, mEntries[ entryIndex ].image );
```
Comment (m-kuhn) [inline src/core/qgsnewsfeedparser.cpp]:
```suggestion
      emit imageFetched( entry.key, mEntries[ entryIndex ].image );
```

is this required for some reason or did you change to the pythonista first camp?

**Review by m-kuhn — APPROVED** [review]
Love it and will not leave a comment for the layout !

File: src/core/qgsnewsfeedparser.cpp (hunk 6 above)
Comment (nyalldawson) [reply src/core/qgsnewsfeedparser.cpp]: It's needed inside lambdas where "this" is captured.

File: src/app/qgswelcomepage.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/app/qgswelcomepage.cpp]: I've added this setting to the QGIS options dialog, and updated the string accordingly

File: src/core/qgsnewsfeedparser.h (hunk 4 above)
Comment (nyalldawson) [reply src/core/qgsnewsfeedparser.h]: We lose Python class docs if I change this to a struct though...

File: src/core/qgsnewsfeedparser.h (hunk 5 above)
Comment (nyalldawson) [reply src/core/qgsnewsfeedparser.h]: Just kinda seems "slotty" to me - e.g. fetching after a timeout, fetching after a button press, ... (I've kept this particular class generic, in case anyone wants to re-use outside of the welcome page)

File: src/core/qgsnewsfeedparser.h (hunk 4 above)
Comment (m-kuhn) [reply src/core/qgsnewsfeedparser.h]: Seriously? Ouch!
